File: internal/icebergmeta/metadata.go
// Package icebergmeta builds and evolves Apache Iceberg table metadata — the
// metadata.json document an Iceberg catalog points a table at.
//
// Overcast writes that document itself where AWS does: S3 Tables' CreateTable
// with a schema, and the Iceberg REST catalog's create and commit (and, later,
// Glue's OpenTableFormatInput). It hand-rolls the format-version 1 and 2 model
// rather than depending on apache/iceberg-go, which would pull arrow-go into
// every binary. Only metadata is handled here; data and manifest files are
// always the engine's or the client's.
//
// The API is Decision 4's in docs/plans/athena-s3tables-iceberg.md: New builds
// a table's first metadata, Parse reads a metadata file, and Commit applies an
// Iceberg REST commit — its requirements, then its updates — to it. The
// package does no I/O and reads no clock: callers pass now.
//
// Spec: https://iceberg.apache.org/spec/#table-metadata-fields
package icebergmeta

import (
	"errors"
	"fmt"
	"maps"
	"regexp"
	"slices"
	"strconv"
	"time"
)

// FormatVersion is the newest Iceberg table format this package writes, and
// the one a new table gets unless its properties ask for version 1.
const FormatVersion = 2

// minFormatVersion is the oldest format this package reads and writes.
const minFormatVersion = 1

// Identifiers the spec gives an initial table.
const (
	// InitialSchemaID is the id of a new table's only schema.
	InitialSchemaID = 0
	// InitialSpecID is the id of a new table's partition spec.
	InitialSpecID = 0
	// UnsortedOrderID is the id the spec reserves for the unsorted order.
	UnsortedOrderID = 0
	// firstPartitionFieldID is where partition field ids start; the reference
	// implementation reports 999 as the last id of a spec with no fields.
	firstPartitionFieldID = 1000
	// noSnapshot is current-snapshot-id for a table without one, as the
	// reference implementation writes it.
	noSnapshot = -1
)

// Table properties this package reads.
const (
	// propFormatVersion chooses a new table's format version. It is reserved:
	// consumed at creation and never stored.
	propFormatVersion = "format-version"
	// propPreviousVersionsMax caps the metadata log; the spec's default is 100.
	propPreviousVersionsMax    = "write.metadata.previous-versions-max"
	defaultPreviousVersionsMax = 100
)

// Schema is an Iceberg struct schema.
type Schema struct {
	Type               string  `json:"type"`
	SchemaID           int     `json:"schema-id"`
	IdentifierFieldIDs []int   `json:"identifier-field-ids,omitempty"`
	Fields             []Field `json:"fields"`
}

// PartitionField is one field of a partition spec.
type PartitionField struct {
	SourceID  int    `json:"source-id"`
	FieldID   int    `json:"field-id"`
	Name      string `json:"name"`
	Transform string `json:"transform"`
}

// PartitionSpec is an Iceberg partition spec.
type PartitionSpec struct {
	SpecID int              `json:"spec-id"`
	Fields []PartitionField `json:"fields"`
}

// SortField is one field of a sort order.
type SortField struct {
	SourceID  int    `json:"source-id"`
	Transform string `json:"transform"`
	Direction string `json:"direction"`
	NullOrder string `json:"null-order"`
}

// SortOrder is an Iceberg sort order.
type SortOrder struct {
	OrderID int         `json:"order-id"`
	Fields  []SortField `json:"fields"`
}

// Snapshot is one snapshot of a table. A version 1 snapshot may list its
// manifests inline instead of naming a manifest list, and has no sequence
// number; the reference implementation writes sequence-number only when it is
// above zero, and so does this package.
type Snapshot struct {
	SnapshotID       int64             `json:"snapshot-id"`
	ParentSnapshotID *int64            `json:"parent-snapshot-id,omitempty"`
	SequenceNumber   int64             `json:"sequence-number,omitempty"`
	TimestampMS      int64             `json:"timestamp-ms"`
	ManifestList     string            `json:"manifest-list,omitempty"`
	Manifests        []string          `json:"manifests,omitempty"`
	Summary          map[string]string `json:"summary,omitempty"`
	SchemaID         *int              `json:"schema-id,omitempty"`
}

// Ref types and the branch every table's current snapshot is on.
const (
	RefBranch  = "branch"
	RefTag     = "tag"
	MainBranch = "main"
)

// SnapshotRef is a named branch or tag.
type SnapshotRef struct {
	SnapshotID         int64  `json:"snapshot-id"`
	Type               string `json:"type"`
	MinSnapshotsToKeep *int   `json:"min-snapshots-to-keep,omitempty"`
	MaxSnapshotAgeMS   *int64 `json:"max-snapshot-age-ms,omitempty"`
	MaxRefAgeMS        *int64 `json:"max-ref-age-ms,omitempty"`
}

// SnapshotLogEntry records when a snapshot became the table's current one.
type SnapshotLogEntry struct {
	TimestampMS int64 `json:"timestamp-ms"`
	SnapshotID  int64 `json:"snapshot-id"`
}

// MetadataLogEntry records a previous metadata file of the table.
type MetadataLogEntry struct {
	TimestampMS  int64  `json:"timestamp-ms"`
	MetadataFile string `json:"metadata-file"`
}

// Metadata is a table metadata document. Field order follows the spec's table
// so the written JSON reads the way the spec does. Schema and PartitionSpec
// are the version 1 fields, written only for a version 1 table.
// CurrentSnapshotID is -1 for a table with no current snapshot, as the
// reference implementation writes it.
type Metadata struct {
	FormatVersion       int                    `json:"format-version"`
	TableUUID           string                 `json:"table-uuid"`
	Location            string                 `json:"location"`
	LastSequenceNumber  int64                  `json:"last-sequence-number"`
	LastUpdatedMS       int64                  `json:"last-updated-ms"`
	LastColumnID        int                    `json:"last-column-id"`
	Schema              *Schema                `json:"schema,omitempty"`
	CurrentSchemaID     int                    `json:"current-schema-id"`
	Schemas             []Schema               `json:"schemas"`
	PartitionSpec       *[]PartitionField      `json:"partition-spec,omitempty"`
	DefaultSpecID       int                    `json:"default-spec-id"`
	PartitionSpecs      []PartitionSpec        `json:"partition-specs"`
	LastPartitionID     int                    `json:"last-partition-id"`
	DefaultSortOrderID  int                    `json:"default-sort-order-id"`
	SortOrders          []SortOrder            `json:"sort-orders"`
	Properties          map[string]string      `json:"properties"`
	CurrentSnapshotID   int64                  `json:"current-snapshot-id"`
	Snapshots           []Snapshot             `json:"snapshots"`
	SnapshotLog         []SnapshotLogEntry     `json:"snapshot-log"`
	MetadataLog         []MetadataLogEntry     `json:"metadata-log"`
	Refs                map[string]SnapshotRef `json:"refs"`
	Statistics          []StatisticsFile       `json:"statistics"`
	PartitionStatistics []StatisticsFile       `json:"partition-statistics"`
}

// CreateSpec describes a new table in Iceberg's own terms. Fields carry the
// caller's field ids, which partition fields, sort fields and
// IdentifierFieldIDs refer to columns by; New then assigns fresh ids from 1,
// as Iceberg does for every new table whatever the caller proposed, and
// rewrites those references to match. A partition field's FieldID is optional
// (zero means "assign one").
type CreateSpec struct {
	TableUUID          string
	Location           string
	Fields             []Field
	IdentifierFieldIDs []int
	PartitionFields    []PartitionField
	SortOrderID        int
	SortFields         []SortField
	Properties         map[string]string
}

// ErrInvalid is matched (errors.Is) by every error for a table definition or
// an update the spec does not allow. The error's message is written for the
// API caller.
var ErrInvalid = errors.New("icebergmeta: invalid table definition")

// ErrRequirementFailed is matched (errors.Is) by the error Commit returns when
// the table no longer meets one of the commit's requirements — another writer
// committed first. The message is written for the API caller.
var ErrRequirementFailed = errors.New("icebergmeta: commit requirement failed")

type metaError struct {
	reason string
	kind   error
}

func (e metaError) Error() string        { return e.reason }
func (e metaError) Is(target error) bool { return target == e.kind }

func invalid(format string, args ...any) error {
	return metaError{reason: fmt.Sprintf(format, args...), kind: ErrInvalid}
}

func requirementFailed(format string, args ...any) error {
	return metaError{reason: fmt.Sprintf(format, args...), kind: ErrRequirementFailed}
}

// New builds the metadata of a table that has no snapshots yet, last updated
// at now: one schema (id 0) with fresh field ids, the given partition spec as
// spec 0 (unpartitioned when empty) and the given sort order (unsorted, id 0,
// when empty). The reserved format-version property picks version 1 or 2 and
// is not stored.
func New(spec CreateSpec, now time.Time) (*Metadata, error) {
	if spec.TableUUID == "" || spec.Location == "" {
		return nil, invalid("a table needs a UUID and a location")
	}
	props := maps.Clone(spec.Properties)
	if props == nil {
		props = map[string]string{}
	}
	version, err := createFormatVersion(props)
	if err != nil {
		return nil, err
	}
	schema, mapping, err := newSchema(spec.Fields, spec.IdentifierFieldIDs)
	if err != nil {
		return nil, err
	}
	partitionFields, lastPartitionID, err := bindPartitionFields(remapSources(spec.PartitionFields, mapping), schema, firstPartitionFieldID-1)
	if err != nil {
		return nil, err
	}
	orders, defaultOrder, err := newSortOrders(spec.SortOrderID, remapSortSources(spec.SortFields, mapping), schema)
	if err != nil {
		return nil, err
	}
	m := &Metadata{
		FormatVersion:      version,
		TableUUID:          spec.TableUUID,
		Location:           spec.Location,
		LastUpdatedMS:      now.UnixMilli(),
		LastColumnID:       schema.highestFieldID(),
		CurrentSchemaID:    InitialSchemaID,
		Schemas:            []Schema{schema},
		DefaultSpecID:      InitialSpecID,
		PartitionSpecs:     []PartitionSpec{{SpecID: InitialSpecID, Fields: partitionFields}},
		LastPartitionID:    lastPartitionID,
		DefaultSortOrderID: defaultOrder,
		SortOrders:         orders,
		Properties:         props,
		CurrentSnapshotID:  noSnapshot,
	}
	m.fillEmpty()
	m.writeVersionFields()
	return m, nil
}

// createFormatVersion consumes the reserved format-version property.
func createFormatVersion(props map[string]string) (int, error) {
	raw, ok := props[propFormatVersion]
	if !ok {
		return FormatVersion, nil
	}
	delete(props, propFormatVersion)
	v, err := strconv.Atoi(raw)
	if err != nil || v < minFormatVersion || v > FormatVersion {
		return 0, invalid("Unsupported format version %q: this catalog writes versions %d and %d.", raw, minFormatVersion, FormatVersion)
	}
	return v, nil
}

// newSchema is the initial schema: the caller's fields validated and
// renumbered from 1, with the old-to-new id mapping.
func newSchema(fields []Field, identifierIDs []int) (Schema, map[int]int, error) {
	checked, err := validateSchema(Schema{Fields: fields, IdentifierFieldIDs: identifierIDs})
	if err != nil {
		return Schema{}, nil, err
	}
	fresh, mapping := freshIDs(checked.Fields)
	schema := Schema{Type: "struct", SchemaID: InitialSchemaID, Fields: fresh}
	for _, id := range identifierIDs {
		schema.IdentifierFieldIDs = append(schema.IdentifierFieldIDs, mapping[id])
	}
	return schema, mapping, nil
}

// remapSources rewrites partition source ids through the fresh-id mapping. An
// id the caller's schema never declared becomes -1, which validation refuses.
func remapSources(in []PartitionField, mapping map[int]int) []PartitionField {
	out := slices.Clone(in)
	for i := range out {
		out[i].SourceID = remapped(mapping, out[i].SourceID)
	}
	return out
}

func remapSortSources(in []SortField, mapping map[int]int) []SortField {
	out := slices.Clone(in)
	for i := range out {
		out[i].SourceID = remapped(mapping, out[i].SourceID)
	}
	return out
}

func remapped(mapping map[int]int, id int) int {
	if v, ok := mapping[id]; ok {
		return v
	}
	return -1
}

// bindPartitionFields validates a spec's fields against schema and assigns an
// id to each field that has none, above both lastAssigned and the highest
// explicit id, so the two never collide. It returns the fields and the new
// last assigned id.
func bindPartitionFields(in []PartitionField, schema Schema, lastAssigned int) ([]PartitionField, int, error) {
	lastID := lastAssigned
	for _, p := range in {
		lastID = max(lastID, p.FieldID)
	}
	out := make([]PartitionField, 0, len(in))
	ids := map[int]bool{}
	names := map[string]bool{}
	for _, p := range in {
		if !schema.hasFieldID(p.SourceID) {
			return nil, 0, invalid("Partition field %q refers to unknown source id %d.", p.Name, p.SourceID)
		}
		if p.Name == "" || p.Transform == "" {
			return nil, 0, invalid("A partition field needs a name and a transform.")
		}
		if p.FieldID == 0 {
			lastID++
			p.FieldID = lastID
		}
		if p.FieldID < firstPartitionFieldID || ids[p.FieldID] {
			return nil, 0, invalid("Partition field id %d is invalid or used more than once.", p.FieldID)
		}
		if names[p.Name] {
			return nil, 0, invalid("Partition field name %q is used more than once.", p.Name)
		}
		ids[p.FieldID], names[p.Name] = true, true
		out = append(out, p)
	}
	return out, lastID, nil
}

// newSortOrders is the unsorted order plus, when fields are given, the order
// orderID over them, with the id of the default order.
func newSortOrders(orderID int, in []SortField, schema Schema) ([]SortOrder, int, error) {
	orders := []SortOrder{unsortedOrder()}
	if len(in) == 0 {
		return orders, UnsortedOrderID, nil
	}
	if orderID == UnsortedOrderID {
		return nil, 0, invalid("Sort order id 0 is reserved for the unsorted order.")
	}
	if err := validateSortFields(in, schema); err != nil {
		return nil, 0, err
	}
	return append(orders, SortOrder{OrderID: orderID, Fields: slices.Clone(in)}), orderID, nil
}

func unsortedOrder() SortOrder { return SortOrder{OrderID: UnsortedOrderID, Fields: []SortField{}} }

func validateSortFields(in []SortField, schema Schema) error {
	for _, f := range in {
		if !schema.hasFieldID(f.SourceID) {
			return invalid("Sort field refers to unknown source id %d.", f.SourceID)
		}
		if f.Direction != "asc" && f.Direction != "desc" {
			return invalid("Sort direction %q must be asc or desc.", f.Direction)
		}
		if f.NullOrder != "nulls-first" && f.NullOrder != "nulls-last" {
			return invalid("Sort null-order %q must be nulls-first or nulls-last.", f.NullOrder)
		}
		if f.Transform == "" {
			return invalid("A sort field needs a transform.")
		}
	}
	return nil
}

// fillEmpty replaces nil collections with empty ones, so the document always
// writes [] and {} rather than null.
func (m *Metadata) fillEmpty() {
	m.Schemas = nonNil(m.Schemas)
	m.PartitionSpecs = nonNil(m.PartitionSpecs)
	m.SortOrders = nonNil(m.SortOrders)
	m.Snapshots = nonNil(m.Snapshots)
	m.SnapshotLog = nonNil(m.SnapshotLog)
	m.MetadataLog = nonNil(m.MetadataLog)
	m.Statistics = nonNil(m.Statistics)
	m.PartitionStatistics = nonNil(m.PartitionStatistics)
	if m.Properties == nil {
		m.Properties = map[string]string{}
	}
	if m.Refs == nil {
		m.Refs = map[string]SnapshotRef{}
	}
	for i := range m.PartitionSpecs {
		m.PartitionSpecs[i].Fields = nonNil(m.PartitionSpecs[i].Fields)
	}
	for i := range m.SortOrders {
		m.SortOrders[i].Fields = nonNil(m.SortOrders[i].Fields)
	}
}

// writeVersionFields sets the version 1 fields a version 1 writer must still
// write — the current schema and the default spec's fields — and clears them
// for version 2.
func (m *Metadata) writeVersionFields() {
	m.Schema, m.PartitionSpec = nil, nil
	if m.FormatVersion != minFormatVersion {
		return
	}
	if s, ok := m.schemaByID(m.CurrentSchemaID); ok {
		m.Schema = &s
	}
	if spec, ok := m.specByID(m.DefaultSpecID); ok {
		fields := spec.Fields
		m.PartitionSpec = &fields
	}
}

func (m *Metadata) schemaByID(id int) (Schema, bool) {
	i := slices.IndexFunc(m.Schemas, func(s Schema) bool { return s.SchemaID == id })
	if i < 0 {
		return Schema{}, false
	}
	return m.Schemas[i], true
}

func (m *Metadata) specByID(id int) (PartitionSpec, bool) {
	i := slices.IndexFunc(m.PartitionSpecs, func(s PartitionSpec) bool { return s.SpecID == id })
	if i < 0 {
		return PartitionSpec{}, false
	}
	return m.PartitionSpecs[i], true
}

func (m *Metadata) hasSortOrder(id int) bool {
	return slices.ContainsFunc(m.SortOrders, func(o SortOrder) bool { return o.OrderID == id })
}

// CurrentSnapshot is the table's current snapshot; ok is false for a table
// with none.
func (m *Metadata) CurrentSnapshot() (Snapshot, bool) {
	i := m.snapshotIndex(m.CurrentSnapshotID)
	if i < 0 {
		return Snapshot{}, false
	}
	return m.Snapshots[i], true
}

func (m *Metadata) snapshotIndex(id int64) int {
	return slices.IndexFunc(m.Snapshots, func(s Snapshot) bool { return s.SnapshotID == id })
}

// metadataFileName is the reference implementation's metadata file name,
// "<version>-<uuid>.metadata.json".
var metadataFileName = regexp.MustCompile(`(?:^|/)(\d+)-[^/]*\.metadata\.json$`)

// MetadataPath is where the version-th metadata file of a table lives,
// relative to the table's location, in the "metadata/<version>-<uuid>.metadata.json"
// form the reference implementation writes.
func MetadataPath(version int, fileUUID string) string {
	return fmt.Sprintf("metadata/%05d-%s.metadata.json", version, fileUUID)
}

// NextMetadataVersion is the version of the metadata file that follows the
// one at location: one more than the version its name carries, or 0 when
// there is no current file or its name carries none.
func NextMetadataVersion(location string) int {
	m := metadataFileName.FindStringSubmatch(location)
	if m == nil {
		return 0
	}
	v, err := strconv.Atoi(m[1])
	if err != nil {
		return 0
	}
	return v + 1
}

File: web/src/components/iceberg/metadata.ts
import { parseJsonKeepingLargeIntegers } from "@/lib/json-text"
import { isRecord } from "@/lib/utils"

/**
 * A typed reader for an Iceberg table metadata file (`*.metadata.json`),
 * shared by everything that shows one: the S3 object preview, the S3 Tables
 * table page and the Glue table page.
 *
 * Tolerant by design. A file written by any engine, any format version, or
 * cut short by a preview window either reads as table metadata or is `null` —
 * a missing optional field becomes an absent property rather than an error,
 * because a viewer that refuses a file shows the developer less than one that
 * shows what it could read.
 *
 * Snapshot ids are random 64-bit integers, far past the 2^53 a JavaScript
 * number holds exactly, so they are read as decimal strings (see
 * `parseJsonKeepingLargeIntegers`).
 *
 * Field names follow the Iceberg table spec's "Table Metadata Fields":
 * https://iceberg.apache.org/spec/#table-metadata-fields
 */

/** A primitive type's spelling (`long`, `decimal(10, 2)`), or a nested type as the file writes it. */
export type IcebergType = string | Record<string, unknown>

export interface IcebergField {
  id: number
  name: string
  required: boolean
  type: IcebergType
  doc?: string
}

export interface IcebergSchema {
  schemaId: number
  fields: IcebergField[]
}

export interface IcebergPartitionField {
  sourceId: number
  fieldId?: number
  name: string
  transform: string
}

export interface IcebergPartitionSpec {
  specId: number
  fields: IcebergPartitionField[]
}

export interface IcebergSnapshot {
  snapshotId: string
  parentSnapshotId?: string
  sequenceNumber?: number
  timestampMs: number
  /** `summary.operation`: `append`, `overwrite`, `replace` or `delete`. */
  operation?: string
  /** The rest of the summary: `added-records`, `total-records`… as the file writes them. */
  summary: Record<string, string>
  manifestList?: string
  schemaId?: number
}

export interface IcebergMetadataLogEntry {
  timestampMs: number
  metadataFile: string
}

export interface IcebergRef {
  name: string
  type: "branch" | "tag"
  snapshotId: string
}

export interface IcebergMetadata {
  formatVersion: number
  tableUuid?: string
  location?: string
  lastUpdatedMs?: number
  lastSequenceNumber?: number
  schemas: IcebergSchema[]
  /** The schema new data is written with: `current-schema-id`'s, or v1's single `schema`. */
  currentSchema?: IcebergSchema
  partitionSpecs: IcebergPartitionSpec[]
  defaultSpec?: IcebergPartitionSpec
  properties: Record<string, string>
  /** Null when the table has no snapshot yet (`-1` or absent — both appear in the wild). */
  currentSnapshotId: string | null
  snapshots: IcebergSnapshot[]
  /** Earlier metadata files of this table, oldest first. Kept up to `write.metadata.previous-versions-max`. */
  metadataLog: IcebergMetadataLogEntry[]
  refs: IcebergRef[]
}

type Json = Record<string, unknown>

/** The objects in a JSON array, or none when `value` is not an array. */
function objectsIn(value: unknown): Json[] {
  return Array.isArray(value) ? value.filter(isRecord) : []
}

function num(value: unknown): number | undefined {
  return typeof value === "number" ? value : undefined
}

function str(value: unknown): string | undefined {
  return typeof value === "string" ? value : undefined
}

/** An id that may arrive as a small number or, past 2^53, as the string the parser kept. */
function id(value: unknown): string | undefined {
  if (typeof value === "string" && /^-?\d+$/.test(value)) return value
  if (typeof value === "number" && Number.isInteger(value)) return String(value)
  return undefined
}

function stringMap(value: unknown): Record<string, string> {
  if (!isRecord(value)) return {}
  return Object.fromEntries(
    Object.entries(value).flatMap(([k, v]) => (typeof v === "string" ? [[k, v]] : [])),
  )
}

/**
 * The file's metadata, or null when `text` is not Iceberg table metadata —
 * including when it is not JSON at all, which a preview window can cause: a
 * metadata file past its limit arrives cut short and cannot be parsed.
 */
export function parseIcebergMetadata(text: string): IcebergMetadata | null {
  let doc: unknown
  try {
    doc = parseJsonKeepingLargeIntegers(text)
  } catch {
    return null
  }
  if (!isRecord(doc)) return null
  const formatVersion = doc["format-version"]
  if (typeof formatVersion !== "number") return null
  if (typeof doc["table-uuid"] !== "string" && typeof doc.location !== "string") return null

  const schemas = readSchemas(doc)
  const partitionSpecs = readPartitionSpecs(doc)
  const currentSnapshotId = id(doc["current-snapshot-id"])
  return {
    formatVersion,
    tableUuid: str(doc["table-uuid"]),
    location: str(doc.location),
    lastUpdatedMs: num(doc["last-updated-ms"]),
    lastSequenceNumber: num(doc["last-sequence-number"]),
    schemas,
    currentSchema: schemas.find((s) => s.schemaId === doc["current-schema-id"]) ?? schemas.at(-1),
    partitionSpecs,
    defaultSpec:
      partitionSpecs.find((s) => s.specId === doc["default-spec-id"]) ?? partitionSpecs.at(-1),
    properties: stringMap(doc.properties),
    currentSnapshotId:
      currentSnapshotId === undefined || currentSnapshotId === "-1" ? null : currentSnapshotId,
    snapshots: objectsIn(doc.snapshots).flatMap(readSnapshot),
    metadataLog: objectsIn(doc["metadata-log"]).flatMap((e) => {
      const file = str(e["metadata-file"])
      const ts = num(e["timestamp-ms"])
      return file && ts !== undefined ? [{ metadataFile: file, timestampMs: ts }] : []
    }),
    refs: readRefs(doc.refs),
  }
}

function readField(f: Json): IcebergField {
  return {
    id: num(f.id) ?? -1,
    name: String(f.name ?? ""),
    required: f.required === true,
    type: typeof f.type === "string" || isRecord(f.type) ? f.type : "unknown",
    doc: str(f.doc),
  }
}

/** v2 and later list every schema; v1 carries a single `schema`. */
function readSchemas(doc: Json): IcebergSchema[] {
  const listed = objectsIn(doc.schemas)
  const raw = listed.length > 0 ? listed : isRecord(doc.schema) ? [doc.schema] : []
  return raw.map((s, index) => ({
    schemaId: num(s["schema-id"]) ?? index,
    fields: objectsIn(s.fields).map(readField),
  }))
}

/** v2 lists every spec; v1 carries the spec inline as `partition-spec`, a bare field list. */
function readPartitionSpecs(doc: Json): IcebergPartitionSpec[] {
  const readFields = (fields: unknown) =>
    objectsIn(fields).map((f) => ({
      sourceId: num(f["source-id"]) ?? -1,
      fieldId: num(f["field-id"]),
      name: String(f.name ?? ""),
      transform: String(f.transform ?? "identity"),
    }))
  const listed = objectsIn(doc["partition-specs"])
  if (listed.length > 0) {
    return listed.map((s, index) => ({
      specId: num(s["spec-id"]) ?? index,
      fields: readFields(s.fields),
    }))
  }
  return Array.isArray(doc["partition-spec"])
    ? [{ specId: 0, fields: readFields(doc["partition-spec"]) }]
    : []
}

function readSnapshot(s: Json): IcebergSnapshot[] {
  const snapshotId = id(s["snapshot-id"])
  const timestampMs = num(s["timestamp-ms"])
  if (snapshotId === undefined || timestampMs === undefined) return []
  const { operation, ...summary } = stringMap(s.summary)
  return [
    {
      snapshotId,
      parentSnapshotId: id(s["parent-snapshot-id"]),
      sequenceNumber: num(s["sequence-number"]),
      timestampMs,
      operation,
      summary,
      manifestList: str(s["manifest-list"]),
      schemaId: num(s["schema-id"]),
    },
  ]
}

function readRefs(value: unknown): IcebergRef[] {
  if (!isRecord(value)) return []
  return Object.entries(value).flatMap(([name, ref]): IcebergRef[] => {
    if (!isRecord(ref)) return []
    const snapshotId = id(ref["snapshot-id"])
    if (snapshotId === undefined) return []
    return [{ name, type: ref.type === "tag" ? "tag" : "branch", snapshotId }]
  })
}

/** Iceberg types as their spec spellings: `long`, `decimal(10, 2)`, `list<string>`, `map<string, long>`, `struct<…>`. */
export function typeName(type: unknown): string {
  if (typeof type === "string") return type
  if (!isRecord(type)) return "unknown"
  switch (type.type) {
    case "list":
      return `list<${typeName(type.element)}>`
    case "map":
      return `map<${typeName(type.key)}, ${typeName(type.value)}>`
    case "struct": {
      const fields = objectsIn(type.fields).map((f) => `${String(f.name)}: ${typeName(f.type)}`)
      return `struct<${fields.join(", ")}>`
    }
    default:
      return typeof type.type === "string" ? type.type : "unknown"
  }
}

/** The struct fields nested inside a type — a struct's own, or a list's or map's struct element. */
export function nestedFields(type: unknown): IcebergField[] {
  if (!isRecord(type)) return []
  if (type.type === "struct") return objectsIn(type.fields).map(readField)
  if (type.type === "list") return nestedFields(type.element)
  if (type.type === "map") return nestedFields(type.value)
  return []
}

/** A field in a schema, with where it sits: `shipping.carrier` is `shipping`'s child at depth 1. */
export interface FieldAtPath {
  field: IcebergField
  path: string
  depth: number
}

/** Every field in `fields`, nested ones after their parent, depth-first in declaration order. */
export function flattenFields(fields: IcebergField[], parent?: FieldAtPath): FieldAtPath[] {
  return fields.flatMap((field) => {
    const here: FieldAtPath = {
      field,
      path: parent ? `${parent.path}.${field.name}` : field.name,
      depth: parent ? parent.depth + 1 : 0,
    }
    return [here, ...flattenFields(nestedFields(field.type), here)]
  })
}

/** A partition field as `transform(column)` — `day(ordered_at)`, `bucket[16](id)`. */
export function partitionLabel(field: IcebergPartitionField, schema?: IcebergSchema): string {
  const source =
    flattenFields(schema?.fields ?? []).find((f) => f.field.id === field.sourceId)?.path ??
    field.name
  return `${field.transform}(${source})`
}
